File: frontend/src/services/driverService.ts
import { api } from "./api";

export type DriverStatus = "ACTIVE" | "INACTIVE";

export type Driver = {
  id: number;
  name: string;
  phone: string;
  licenseNumber: string;
  assignedBus: string;
  status: DriverStatus;
  createdAt: string;
};

export type DriverPayload = {
  name: string;
  phone: string;
  licenseNumber: string;
  assignedBus: string;
  status: DriverStatus;
};

export type DriverMetrics = {
  total: number;
  active: number;
  inactive: number;
};

export async function list(): Promise<Driver[]> {
  const res = await api.get<Driver[]>("/driver");
  return res.data;
}

export async function get(id: number): Promise<Driver> {
  const res = await api.get<Driver>(`/driver/${id}`);
  return res.data;
}

export async function create(payload: DriverPayload): Promise<Driver> {
  const res = await api.post<Driver>("/driver", payload);
  return res.data;
}

export async function update(id: number, payload: DriverPayload): Promise<Driver> {
  const res = await api.put<Driver>(`/driver/${id}`, payload);
  return res.data;
}

export async function remove(id: number): Promise<void> {
  await api.delete(`/driver/${id}`);
}

export async function metrics(): Promise<DriverMetrics> {
  const res = await api.get<DriverMetrics>("/driver/metrics");
  return res.data;
}

File: driver-service/src/main/java/com/smarttransport/driverservice/service/DriverService.java
package com.smarttransport.driverservice.service;

import com.smarttransport.driverservice.domain.Driver;
import com.smarttransport.driverservice.domain.DriverStatus;
import com.smarttransport.driverservice.dto.DriverMetrics;
import com.smarttransport.driverservice.dto.DriverRequest;
import com.smarttransport.driverservice.repository.DriverRepository;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Objects;

@Service
public class DriverService {

    private final DriverRepository repository;

    public DriverService(DriverRepository repository) {
        this.repository = repository;
    }

    public List<Driver> list() {
        return repository.findAll();
    }

    public Driver get(@NonNull Long id) {
        Long safeId = requireId(id);
        return repository.findById(safeId).orElseThrow(() -> new NotFoundException("Driver not found"));
    }

    @Transactional
    public Driver create(DriverRequest request) {
        Driver driver = new Driver(request.getFullName(), request.getPhone(), request.getLicenseNumber(), request.getStatus());
        try {
            return repository.save(driver);
        } catch (DataIntegrityViolationException e) {
            throw new BadRequestException("License number already exists");
        }
    }

    @Transactional
    public Driver update(@NonNull Long id, DriverRequest request) {
        Long safeId = requireId(id);
        Driver existing = get(safeId);
        existing.setFullName(request.getFullName());
        existing.setPhone(request.getPhone());
        existing.setLicenseNumber(request.getLicenseNumber());
        existing.setStatus(request.getStatus());
        try {
            return repository.save(existing);
        } catch (DataIntegrityViolationException e) {
            throw new BadRequestException("License number already exists");
        }
    }

    @Transactional
    public void delete(@NonNull Long id) {
        Long safeId = requireId(id);
        if (!repository.existsById(safeId)) throw new NotFoundException("Driver not found");
        repository.deleteById(safeId);
    }

    public DriverMetrics metrics() {
        long total = repository.count();
        long active = repository.countByStatus(DriverStatus.ACTIVE);
        long inactive = repository.countByStatus(DriverStatus.INACTIVE);
        return new DriverMetrics(total, active, inactive);
    }

    private static @NonNull Long requireId(@NonNull Long id) {
        return Objects.requireNonNull(id, "Driver id is required");
    }
}
